packages/next/src/build/webpack/plugins/slow-module-detection-plugin.ts

```diff
@@ -0,0 +1,330 @@
+import type { Compiler, Module, Compilation } from 'webpack'
+import type { CompilerNameValues } from '../../../shared/lib/constants'
+
+/**
+ * Plugin that detects and reports modules that take a long time to build.
+ * Helps identify performance bottlenecks in the build process by:
+ * 1. Tracking build time for each module
+ * 2. Building a dependency tree of slow modules
+ * 3. Generating a visual report with timing information
+ */
+
+// Configuration constants
+const PLUGIN_NAME = 'SlowModuleDetectionPlugin'
+
+// Console output styling - using ANSI escape codes for colors
+const ConsoleColors = {
+  BOLD_BLUE: '\x1b[1;34m',
+  BRIGHT_YELLOW: '\x1b[93m',
+  BOLD_GREEN: '\x1b[1;32m',
+  RESET: '\x1b[0m',
+} as const
```

Comment:
we have `~/lib/picocolors` for this. This ensures coloring is not use if not desired/supported.